### PR TITLE
SONARJAVA-2300 fix duplicated messages in flow

### PR DESCRIPTION
**Pull request by saberduck**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java [hunk 1]
```diff
@@ -354,10 +354,18 @@ private boolean isMethodParameter(Symbol symbol) {
     }
 
     private List<JavaFileScannerContext.Location> flowFromLearnedConstraints(ExplodedGraph.Edge edge, Set<LearnedConstraint> learnedConstraints) {
-      return learnedConstraints.stream()
+      // FIXME SONARJAVA-2303 temporary workaround to avoid duplicated messages when constraint is reported on relational SV and also on its operand
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1134)

File: java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java [hunk 2]
```diff
@@ -354,10 +354,18 @@ private boolean isMethodParameter(Symbol symbol) {
...
+      List<JavaFileScannerContext.Location> result = new ArrayList<>();
+      Set<String> messages = new HashSet<>();
+      learnedConstraints.stream()
         .map(lc -> learnedConstraintFlow(lc, edge))
         .flatMap(List::stream)
-        .collect(Collectors.toList());
+        .forEach(location -> {
+          if (!location.syntaxNode.is(Tree.Kind.EQUAL_TO, Tree.Kind.NOT_EQUAL_TO) || messages.add(location.msg)) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java]: MAJOR Unknown method [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=java-debugging-rules%3AUnknownMethodInvocation)
MAJOR Unknown method [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=java-debugging-rules%3AUnknownMethodInvocation)

File: java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java [hunk 3]
```diff
@@ -354,10 +354,18 @@ private boolean isMethodParameter(Symbol symbol) {
...
+            result.add(location);
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java]: MAJOR Unknown method [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=java-debugging-rules%3AUnknownMethodInvocation)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 4 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java (hunk 2 above)
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java]: Mmmh, I'm not sure to follow you here. What is bothering me is having two times the same message on the same tree (so a duplicate of `Location`). So why considering the kind of the tree and only the message value? It's probably way simpler to rely on equality of locations to filter out, with a `LinkedHashSet` for instance.

File: java-frontend/src/test/files/se/FlowComputation.java [hunk 4]
```diff
@@ -48,5 +48,13 @@ void invocation_target(Object a) {
     }
   }
 
+  void test(Object a, Object b) {
+    if (a == b) { // flow@catof
```
Comment (m-g-sonar) [inline java-frontend/src/test/files/se/FlowComputation.java]: Don't know why you added this test without the flow message.  To me it already does not have two messages.

File: java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java (hunk 2 above)
Comment (saberduck) [inline java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java]: You're right, this is non-sense. I was in the wrong mindset. Reimplemented with LinkedHashSet

File: java-frontend/src/test/files/se/FlowComputation.java (hunk 4 above)
Comment (saberduck) [inline java-frontend/src/test/files/se/FlowComputation.java]: Sorry, this is not relevant to this ticket and should be indeed part of SONARJAVA-2303, I removed it from this PR

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java (hunk 2 above)
Comment (saberduck) [inline java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java]: In fact using `distinct` on the stream is perhaps even better

File: java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java [hunk 5]
```diff
@@ -354,9 +354,11 @@ private boolean isMethodParameter(Symbol symbol) {
     }
 
     private List<JavaFileScannerContext.Location> flowFromLearnedConstraints(ExplodedGraph.Edge edge, Set<LearnedConstraint> learnedConstraints) {
+      // FIXME SONARJAVA-2303 calling distinct is temporary workaround to avoid duplicated messages when constraint is reported on relational SV and also on its operand
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1134)

**Pull request merged**